### PR TITLE
Improve loose comparison on union type

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/8800

### CODE REVIEW

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Would ve great to do it with IntersectionType too.

I suspect the implementations should resemble TrinaryLogic::minMax and extremeIdentity so I want to compare them with that.

**Comment by staabm** [comment]
I think its pretty neat now :)

edit: just realized I need to fix the implementations in the accessory types

File: tests/PHPStan/Analyser/nsrt/loose-comparisons.php [hunk 1]
```diff
@@ -653,4 +653,85 @@ public function sayInt(
...
+		if ($upper != '') {
+			assertType('false', '' == $upper);
+		}
+		assertType('bool', '0' == $upper);
+		assertType('bool', 'a' == $upper); // should be false
+		assertType('bool', 'abc' == $upper); // should be false
+		assertType('bool', 'aBc' == $upper);
+		assertType('bool', strtoupper($s) == $upper);
+		assertType('bool', strtolower($s) == $upper); // should be false
+		assertType('bool', $upper == $lower); // should be false
```
Comment (staabm) [inline tests/PHPStan/Analyser/nsrt/loose-comparisons.php]: @VincentLanglet are you interessted in implementing `looseCompare()` in upper/lower-case string type? :)

I tried but can't get my head arround this atm

**Comment by staabm** [comment]
added tests for non-empty-array and non-empty-string to show the intersection-type loose compare is working

File: tests/PHPStan/Analyser/nsrt/loose-comparisons.php (hunk 1 above)
Comment (VincentLanglet) [reply tests/PHPStan/Analyser/nsrt/loose-comparisons.php]: The tests are not true, I edited with comment 
```
		assertType('bool', 'aBc' == $upper); // => Should be false no ?
		
		assertType('bool', strtolower($s) == $upper); // should be false => NO, it can be true with `$s = '&!<42' or an empty string
		
		assertType('bool', $upper == $lower); // should be false => NO, it can be true with `$upper = ''` or `$upper = '&!<42'`
		
		assertType('bool', 'AbC' == $lower); // => Should be false no ?
		
		assertType('bool', strtoupper($s) == $lower); // should be false => NO, it can be true with `$s = '&!<42' or an empty string
		
		assertType('bool', $lower == $upper); // should be false => NO, it can be true with `$lower = ''` or `$lower = '&!<42'`
```

So maybe it's the reason of your difficulties ?

For the implementation I would go with
https://github.com/phpstan/phpstan-src/commit/ff999299810056a8cb419f818a0bb85e0697b758

I had to comment the `if` condition because there is a bug, you might know how to fix
```
// Here PHPStan consider $lower is `lowercase`
if ($lower != '') {
     assertType('false', '' == $lower);
     // Here PHPStan consider $lower is `lowercase&non-empty-string`
}
// Here PHPStan consider $lower is `string`
```

Comment (VincentLanglet) [reply tests/PHPStan/Analyser/nsrt/loose-comparisons.php]: See https://phpstan.org/r/5b9fe2de-f827-4e3b-99ac-9d8b62dca596 @staabm 

So you might consider moving the assertion to avoid this bug https://github.com/phpstan/phpstan-src/commit/36bda290d1487a8705fc802e11cb28e2bfb78964

Comment (staabm) [reply tests/PHPStan/Analyser/nsrt/loose-comparisons.php]: thanks for looking into it. I have added your proposal as a separate commit.

File: src/Type/Accessory/AccessoryLowercaseStringType.php [hunk 2]
```diff
@@ -326,6 +327,10 @@ public function isScalar(): TrinaryLogic
 
 	public function looseCompare(Type $type, PhpVersion $phpVersion): BooleanType
 	{
+		if ($type->isString()->yes() && $type->isLowercaseString()->no()) {
```
Comment (mvorisek) [inline src/Type/Accessory/AccessoryLowercaseStringType.php]: Does https://3v4l.org/tEobn test pass (and vice versa)?

Comment (staabm) [reply src/Type/Accessory/AccessoryLowercaseStringType.php]: Sounds like a very edge case which should be reported as a new issue.

I don't expect it works with this PR

Comment (VincentLanglet) [reply src/Type/Accessory/AccessoryLowercaseStringType.php]: No, it's considered as false with the current implementation.

@staabm I think you'll need to add
```
&& ($type->isNumericString()->no() || $this->isNumericString()->no())
```
condition.

Comment (mvorisek) [reply src/Type/Accessory/AccessoryLowercaseStringType.php:330]: It works already https://phpstan.org/r/3b02891b-463b-4d09-b845-428b10f44547 somehow ;-)

Comment (mvorisek) [reply src/Type/Accessory/AccessoryLowercaseStringType.php]: I do not see any issue yet - https://phpstan.org/r/745f421e-ae6f-46f1-aa32-d2fe28d88b3f . But this PR will probably break it.

Comment (VincentLanglet) [reply src/Type/Accessory/AccessoryLowercaseStringType.php]: You can cherry pick https://github.com/phpstan/phpstan-src/commit/e11991f2dffbf56f2f8c8be9652454547285e6b3 staabm

Comment (staabm) [reply src/Type/Accessory/AccessoryLowercaseStringType.php]: cherry picked and added a few more tests, thanks

File: src/TrinaryLogic.php [hunk 3]
```diff
@@ -47,6 +47,18 @@ public static function createFromBoolean(bool $value): self
 		return self::$registry[$yesNo] ??= new self($yesNo);
 	}
 
+	public static function createFromBooleanType(BooleanType $type): self
```
Comment (ondrejmirtes) [inline src/TrinaryLogic.php]: You had the same idea, nice! But I'd actually like `BooleanType::toTrinaryLogic()` more :)

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/TrinaryLogic.php (hunk 3 above)
Comment (staabm) [reply src/TrinaryLogic.php]: hehe I see. done

**Comment by ondrejmirtes** [comment]
Thank you!
